### DISCUSSION TITLE
How to create your own auth.x function?

### DISCUSSION THREAD

**Discussion by enricoschaaf** [supabase/supabase#661] [Questions]
I watched this video from supabase https://www.youtube.com/watch?v=0LvCOlELs5U&t=812s and tried to write my write my own auth.roles() function thats very similar to the auth.role() function exept that I "supercharged" my JWT Token so it includes an roles array.

I wrote this function

```PLpgSQL
create function auth.roles() returns text[] as $$
    select nullif( array( select * from json_array_elements_text (current_setting('request.jwt.claim.roles', true)::json)), '{}')::text[];
$$ language sql stable;
```

that should return an array of roles, I tested it as a rpc.

But when I want to use it like this as a policy it just does not work.

```PLpgSQL
create policy "Allow team members to view registrations" on public.registrations
for select using ('TEAM_MEMBER_' || year = any (auth.roles()));
```

Can anyone explain to me why that is? if I use an array of strings instead of auth.roles() it works but my function should return an array of strings or am I missing something? When I call it via bash as a rpc I get ["TEAM_MEMBER_2021"] which is the currently the only role attached to the user.

**Comment by steve-chavez** [comment]
One way you could debug this yourself is by doing a [RAISE EXCEPTION](https://www.postgresql.org/docs/13/plpgsql-errors-and-messages.html) inside your function, the error will be reported through the REST API. That way you can check the `auth.roles` function is giving the right value when inside the policy.

**Reply by enricoschaaf** [reply]
Could not figure out how to to the raise exception thing, but I think I know why it doesnt work but I have no idea how to fix it.

When I run `current_setting('request.jwt.claim.roles', true)` on its own it returns '[\\"TEAM_MEMBER_2021\\"]'

the function then looks like this
```SQL
create or replace function auth.roles() returns text[] as $$
    select nullif( array ( select * from json_array_elements_text ('[\"TEAM_MEMBER_2021\"]') ), '{}')::text[]
$$ language sql stable;
```

The problem is that [\\"TEAM_MEMBER_2021\\"] is invalid json for json_array_elements_text it expects ["TEAM_MEMBER_2021"] but I have no idea how to fix that. 😑

**Reply by enricoschaaf** [reply]
Seems like I was wrong because this works

```SQL
create or replace function auth.roles() returns text[] as $$
    select nullif ( string_to_array( 'TEAM_MEMBER_2021,TEAM_MEMBER_2021', ','), '{}')::text[]
$$ language sql stable;

create policy "Allow team members to view registrations" on public.registrations
for select using ('authenticated' = any (auth.roles()));

```

but this does not, even though I changed roles the the exact same string, so there must be something going on with current_setting what I missed.

```SQL
create or replace function auth.roles() returns text[] as $$
    select nullif ( string_to_array( current_setting('request.jwt.claim.roles', true), ','), '{}')::text[]
$$ language sql stable;

create policy "Allow team members to view registrations" on public.registrations
for select using ('authenticated' = any (auth.roles()));
```

**Reply by enricoschaaf** [reply]
Okey the problem is simply this

{
  hint: null,
  details: null,
  code: '42704',
  message: 'unrecognized configuration parameter "request.jwt.claim.roles"'
}

If you set current_setting to false it show me this error, should have read it before.
Would it be possible to simple expose all the jwt properties so stuff like this would work? It works when executed as a rpc but not inside a policy. @kiwicopple

**Reply by kiwicopple** [reply]
> It works when executed as a rpc but not inside a policy.

 Good catch. 

https://www.postgresql.org/docs/current/sql-createpolicy.html

> Since policy expressions are added to the user's query directly, they will be run with the rights of the user running the overall query.

It seems that the `auth.roles()` function works because it's running with elevated privileges and therefore has access to the request headers (right @steve-chavez ?). In this case, it seems that it will only work with a function, but you could always create a generic function which returns all the headers (rather than just the `roles()`)

**Reply by steve-chavez** [reply]
> It seems that the auth.roles() function works because it's running with elevated privileges and therefore has access to the request headers

@kiwicopple The request gucs like `request.jwt.claim` don't require privileges, they're transaction scoped and can be accessed by any role.

@enricoschaaf I'm trying to reproduce your example. One question, how are you calling the `auth.roles` function? Since we don't expose the `auth` schema, you shouldn't be able to call it as an rpc.

Also, from your ` 'unrecognized configuration parameter "request.jwt.claim.roles"'` error, it seems the JWT you used doesn't include the `roles` claim. How are you generating the JWT?

**Reply by steve-chavez** [reply]
@enricoschaaf I've just tried your example with the `auth.roles()` inside a policy and it works good. So it's not a privilege issue.

I think the problem must lie in how you generated the JWT. What I've done is to paste the `anon` jwt on `/settings/api` on https://jwt.io/  and add the `roles` claim key manually. Then make an API call with this new JWT.

**Comment by technopahadi** [comment]
Super late but for anybody trying to figure the following out:

> Would it be possible to simple expose all the jwt properties so stuff like this would work? 

This will work (due to PostgREST) (`request.jwt.claims` instead of `claim` [singular]):

```
current_setting('request.jwt.claims', false)::json->>'PROPERTY');
```
